### ISSUE TITLE
Lua API: initialize callback of register_storage crashes when returning empty array

### ISSUE THREAD

**Issue by markusb** [darktable-org/darktable#14803]
### Describe the bug

I'm using the register_storage api in a lua script.

With register_storage I define an 'initilize' subroutine getting called before processing starts.
When I determine that I do not want to process any image, I exit the subroutine using an empty array to signal there are no images to be processed. This is described in the API manual. https://docs.darktable.org/lua/stable/lua.api.manual/darktable/darktable.register_storage/

When I do that Darktable crashes immediately with this message.

Magick: caught exception 0xC0000005 "Access violation"...

### Steps to reproduce

1. Install darktable 4.4.0
2. Install the lua scripts
3. Start the enfuseAdvanced script with script manager
4. In the Export module select 'DRI or DIFF image' (this will select the enfuseAdvanced script)
5. Select a single image
6. Press 'Export'
    -> Darktable crashes

### Expected behavior

Properly process the empty table as described in the API Manual

### Logfile | Screenshot | Screencast

Magick: caught exception 0xC0000005 "Access violation"...'git' is not recognized as an internal or external command, operable program or batch file.

### Where did you install darktable from?

darktable.org

### darktable version

4.4.0

### What OS are you using?

Windows

### What is the version of your OS?

Windows 10

### Describe your system?

Lenovo T16 Laptop

### Are you using OpenCL GPU in darktable?

No

### Please provide additional context if applicable. You can attach files too, but might need to rename to .txt or .zip

I can supply a simple piece of code to show the issue, but as I found the same problem in a piece of contrib code I prefer not to. This excludes a problem with my code7script ans I can reproduce the issue with a script from the official repository.

**Comment by wpferguson** [comment]
Which script in the repository?

**Comment by markusb** [comment]
The enfuseAdvanced script.

**Comment by wpferguson** [comment]
When I run the script and trigger the bug, darktable just exits.  There is no backtrace, stack dump, error message, etc.  It is an abnormal ending, since the darktablerc file doesn't have the latest changes written to it.

**Comment by markusb** [comment]
I just get the message Magick: caught exception 0xC0000005 "Access violation".

But there is clearly something broken.

**Comment by wpferguson** [comment]
> But there is clearly something broken.

Agreed, but darktable isn't being very cooperative in telling me where the problem is.  I ran it under a debugger last night, but again didn't get anything useful.  The last time I used gdb (the debugger) in any meaningful way was decades ago, so I need to spruce up my gdb skills and try again.

**Comment by markusb** [comment]
I'm agraid I can not really help here. Never dabbled in Darktable code myself. Also, looks this is a Windows specific problem and I don't even know where to start.

**Comment by wpferguson** [comment]
I did a little more testing on this...

When you run the script and select only one image, the initialize callback detects that there aren't enough images, so it sets a code in the extra data that is returned.  When the next callback is started (main) it checks for the error code and prints a message, then returns, and at that point the script should exit.

When testing the script, I saw the error message, which meant the the initialize callback completed and returned.  Now I can examine what everything is set to and see if I can find the issue.

**Comment by markusb** [comment]
Thanks !

**Comment by koto** [comment]
I can reproduce this, it's not Windows specific, it triggers on MacOS builds too (4.8.0)

```bash
$ mkdir foo
$ /Applications/darktable.app/Contents/MacOS/darktable -d lua --configdir foo --luacmd 'local dt = require "darktable";dt.register_storage("testbug", "test bug", nil, nil, nil, function() return {} end)' image.jpg
darktable 4.8.0
Copyright (C) 2012-2024 Johannes Hanika and other contributors.

Compile options:
  Bit depth              -> 64 bit
  Debug                  -> DISABLED
  SSE2 optimizations     -> DISABLED
  OpenMP                 -> ENABLED
  OpenCL                 -> ENABLED
  Lua                    -> ENABLED  - API version 9.3.0
  Colord                 -> DISABLED
  gPhoto2                -> ENABLED
  GMIC                   -> ENABLED  - Compressed LUTs are supported
  GraphicsMagick         -> DISABLED
  ImageMagick            -> ENABLED
  libavif                -> ENABLED
  libheif                -> ENABLED
  libjxl                 -> ENABLED
  OpenJPEG               -> ENABLED
  OpenEXR                -> ENABLED
  WebP                   -> ENABLED

See https://www.darktable.org/resources/ for detailed documentation.
See https://github.com/darktable-org/darktable/issues/new/choose to report bugs.


(process:10251): GLib-GObject-CRITICAL **: 20:51:46.657: g_object_set: assertion 'G_IS_OBJECT (object)' failed

(darktable:10251): Gtk-CRITICAL **: 20:51:46.869: gtk_window_add_accel_group: assertion 'GTK_IS_WINDOW (window)' failed
     3.4865 [pixelpipe_process CPU] buffer aligment problem: IN=0x124800040 OUT=0x160000000
     3.5264 LUA ERROR : cannot open /Users/koto/foo/luarc: No such file or directory
     3.6173 [pixelpipe_process CPU] buffer aligment problem: IN=0x124800040 OUT=0x30bf40000
     3.6275 [pixelpipe_process CPU] buffer aligment problem: IN=0x15645c040 OUT=0x32bf40000
     3.9275 [pixelpipe_process CPU] buffer aligment problem: IN=0x148f28040 OUT=0x12a088000

(darktable:10251): Pango-WARNING **: 20:51:50.913: couldn't load font "monospace Semi-Condensed Not-Rotated 12", modified variant/weight/stretch as fallback, expect ugly output.
darktable(10251,0x16dc27000) malloc: *** error for object 0x60000121d480: pointer being freed was not allocated
darktable(10251,0x16dc27000) malloc: *** set a breakpoint in malloc_error_break to debug
zsh: abort      /Applications/darktable.app/Contents/MacOS/darktable -d lua --configdir foo
```
To trigger the crash, just select the 'test bug' export storage and start the export.

**Comment by wpferguson** [comment]
Your arguments to register_storage are not correct, the last argument is supposed to be a widget.

The bug described in this issue results from returning an empty array from the initialize function.

**Comment by koto** [comment]
> Your arguments to register_storage are not correct, the last argument is supposed to be a widget.

Yup, it's the same if a widget is passed. Correct, it's all about storage's `initialize` function returning an empty list.

**Comment by wpferguson** [comment]
Fixed by #19309
